### PR TITLE
Parameterise over era

### PR DESCRIPTION
**Pull request by mrBliss**
This updates the dependency on ledger-specs, bringing in a number of changes.

See the individual commits for more details.

### CODE REVIEW

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 1]
```diff
@@ -251,43 +270,54 @@ instance TPraosCrypto c => CanHardFork (CardanoEras c) where
   Translation from Byron to Shelley
 -------------------------------------------------------------------------------}
 
-translateHeaderHashByronToShelley
-  :: forall sc. Crypto sc
+translateHeaderHashByronToShelley ::
+     forall era. (Era era, HASH (Era.Crypto era) ~ Blake2b_256)
```
Comment (nfrisby) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:261]: 👍

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 2]
```diff
@@ -251,43 +270,54 @@ instance TPraosCrypto c => CanHardFork (CardanoEras c) where
...
     GenesisPoint   -> GenesisPoint
     BlockPoint s h -> BlockPoint s (translateHeaderHashByronToShelley h)
 
-
 -- | We use the same hashing algorithm so we can unwrap and rewrap the bytes.
 -- We don't care about the type that is hashed, which will differ going from
 -- Byron to Shelley, we just use the hashes as IDs.
-translateTxIdByronToShelley :: Crypto sc => CC.TxId -> SL.TxId sc
+translateTxIdByronToShelley ::
+     forall era. (Era era, ADDRHASH (Era.Crypto era) ~ Blake2b_224)
```
Comment (nfrisby) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs]: 👍

File: ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs [hunk 3]
```diff
@@ -1,45 +1,50 @@
...
 import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock)
 import           Ouroboros.Consensus.Shelley.Protocol.Crypto (TPraosCrypto)
 
-import           Shelley.Spec.Ledger.Crypto (Crypto (..))
-
--- | A mock replacement for 'TPraosStandardCrypto' that is compatible with
+-- | A mock replacement for 'StandardCrypto' that is compatible with
 -- bootstrapping from Byron
 --
 -- This mocks more components than does
```
Comment (nfrisby) [inline ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs:25]: I think this first sentence is either out-of-date and/or was wrong to begin with. `MockCryptoCompatByron` does _not_ mock more than `MockCrypto` does.

Also, the last item in the commit message 6fdeac5633fe2034080ccbe240c5369fdd604bb6 of says that the generators can be re-used, so the last two sentences are also out-of-date, if I understand correctly.

File: ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs [hunk 4]
```diff
@@ -1,45 +1,50 @@
...
+-- 'Test.Consensus.Shelley.MockCrypto.MockCrypto'. This prevents the
 -- @cardano-ledger-specs@ generators from being re-used. Currently, this is not
 -- an obstacle for example in the Cardano ThreadNet tests.
 --
--- NOTE: The "Ouroboros.Consensus.Cardano.CanHardFork" translation currently
--- assumes that @ADDRHASH@ has the same bit size as Byron address hashes (ie
--- 224); that's why we use 'Blake2b_224' here.
+-- NOTE: The "Ouroboros.Consensus.Cardano.CanHardFork" translation requires that
+-- @ADDRHASH@ has the same bit size as Byron address hashes (ie 224); that's why
+-- we use 'Blake2b_224' here. Similarly, @HASH@ has to hve the same bit size as
```
Comment (nfrisby) [inline ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs:32]: hve -> have

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs [hunk 5]
```diff
@@ -89,18 +90,25 @@ import qualified Ouroboros.Consensus.HardFork.Combinator.Util.Telescope as Teles
...
+type ShelleyEra c = Era.Shelley c
+
 -- | First we had the Byron era and then the Shelley era.
 --
--- We parameterise over the crypto used in the Shelley era: @sc@.
+-- We parameterise over the crypto used in the Shelley era: @c@.
 --
 -- TODO: parameterise ByronBlock over crypto too
-type CardanoEras sc = '[ByronBlock, ShelleyBlock sc]
+type CardanoEras c = '[ByronBlock, ShelleyBlock (ShelleyEra c)]
```
Comment (nfrisby) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs:111]: The zig zag in this definition from era to block back to era is a little confusing. Is there any chance we should rename it to `CardanoBlocks`?

File: ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs [hunk 6]
```diff
@@ -117,31 +111,31 @@ tpraosSlotLength = slotLengthFromSec 2
   CoreNode secrets/etc
 -------------------------------------------------------------------------------}
 
-data CoreNode c = CoreNode {
-      cnGenesisKey  :: !(SignKeyDSIGN (SL.DSIGN c))
-    , cnDelegateKey :: !(SignKeyDSIGN (SL.DSIGN c))
+data CoreNode era = CoreNode {
+      cnGenesisKey  :: !(SL.SignKeyDSIGN era)
```
Comment (nfrisby) [inline ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs:115]: 👍 I'm happy to see this upstream change to `SignKeyDSIGN` et al

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs [hunk 7]
```diff
@@ -32,24 +32,23 @@ import           Test.Util.Nightly
...
-import           Test.Consensus.Shelley.MockCrypto (TPraosMockCrypto)
+import           Test.Consensus.Shelley.MockCrypto (MockShelley)
 import           Test.ThreadNet.TxGen.Shelley
 import           Test.ThreadNet.Util.NodeJoinPlan (trivialNodeJoinPlan)
 import           Test.ThreadNet.Util.NodeRestarts (noRestarts)
 import           Test.ThreadNet.Util.NodeToNodeVersion (genVersion)
 import           Test.ThreadNet.Util.Seed (runGen)
 
-type Crypto = TPraosMockCrypto ShortHash
+type Mock = MockShelley ShortHash
```
Comment (nfrisby) [inline ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs:51]: Perhaps `type Era =` would be more consistent?

**Comment by nfrisby** [comment]
This whole thing looks good to me. I'm not going to Approve, though, because I'm sure @nc6 has a better grip on what recently changed upstream and what was recently moved upstream. I didn't compare the blocks deleted here to the upstream PRs that @mrBliss's comments nicely reference.

Word diff highlighting is a real life saver for this one :D. Nice job on the heavy-lifting, Nick and Thomas!

File: ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs (hunk 3 above)
Comment (mrBliss) [reply ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs:25]: Fixed, thanks.

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs (hunk 5 above)
Comment (mrBliss) [reply ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs:111]: Valid point, but I think `CardanoBlocks` might be confusing w.r.t. to `CardanoBlock`? It also means more breakage. Let's revisit this when we add the next element to the list :slightly_smiling_face:

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs (hunk 7 above)
Comment (mrBliss) [reply ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs:51]: Good idea, done.

**Review by kderme — APPROVED** [review]
LGTM, at least the 2 commits about moving translation and stability window.

**Comment by mrBliss** [comment]
I'll first try propagating this to `cardano-node`, to see if I have forgotten anything.

**Review by nc6 — APPROVED** [review]

File: ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs [hunk 8]
```diff
@@ -1,45 +1,53 @@
...
+-- * The @shelley-spec-ledger@ package currently requires that @'DSIGN' ~
+--   'Ed25519DSIGN' in order to use Byron bootstrap witnesses.
 --
--- NOTE: The @shelley-spec-ledger@ package currently requires that @'DSIGN' ~
--- 'Ed25519DSIGN' in order to use Byron bootstrap witnesses.
-data TPraosMockCryptoCompatByron h
+-- * We can still use mock KES and mock VRF.
+--
+-- Note that many Shelley generators are not instantiated to 'MockShelley' but
+-- are constraint by @'CanMock' era@. 'MockShelleyCompatByron' satisfies this
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/MockCrypto.hs]: Fixed, thanks.

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**